src/main/c/org_broadinstitute_hellbender_utils_bwa_BwaMemIndex.c

```diff
@@ -17,6 +17,30 @@ jint throwIllegalArgumentException(JNIEnv* env, char* message) {
    return (*env)->ThrowNew(env, iaeClass, message);
 }
 
+int jobject_to_mem_pestat_t(JNIEnv* env, jobject in, mem_pestat_t *out) {
+   if (in == NULL) {
+     return 0;
+   }
+   memset(out, 0, sizeof(mem_pestat_t) * 4);
+   jclass pestatClass = (*env)->FindClass(env, "org/broadinstitute/hellbender/utils/bwa/BwaMemPairEndStats");
```

Comment:
These JNI functions can signal failure by return value or by throwing exceptions.  You're not checking either situation.  Also, you really should cache the class and field IDs so that you only need to look them up once.  Also, you're holding a local reference to a Java object for a long time (for the duration of the alignment process).
You can 1) deal with all this complicated JNI crap, or 2) just use a ByteBuffer to communicate data between Java and C the way all the other code in this library does.
If you choose #1, you may find this helpful: https://www.ibm.com/developerworks/library/j-jni/index.html

Reply:
I'm following 1). Not sure if I understand what is the issue on holding referene on the java-object (I guess you are referring to the BwaMemPairEndStats[] argument). Since this call is synchronous such an object cannot be disposed of by the GC until the alignment has concluded, right? What would the damaged done here?

Reply:
This might not be completely academic, because we do have a problem when running locally that local assemblies seem to be serialized, rather than running in parallel.  It remains a mystery as to exactly why, but it seems likely that it has something to do with the JNI.

Reply:
Ok I see.